packages/server-wallet/src/wallet/index.ts

```diff
@@ -210,47 +224,52 @@ const takeActions = async (channels: Bytes32[]): Promise<ExecutionResult> => {
-          const {outgoing, channelResult} = await Store.signState(action.channelId, action, tx);
-          outgoing.map(n => outbox.push(n.notice));
-          channelResults.push(channelResult);
-          return;
+    await knex.transaction(async tx => {
+      const setError = async (e: Error): Promise<void> => {
+        error = e;
+        await tx.rollback(error);
+      };
+      const markChannelAsDone = async (): Promise<any> => {
+        channels.shift() as string;
+      };
+
+      const handleAction = async (action: ProtocolAction): Promise<any> => {
+        switch (action.type) {
+          case 'SignState': {
+            const {outgoing, channelResult} = await Store.signState(action.channelId, action, tx);
+            outgoing.map(n => outbox.push(n.notice));
+            channelResults.push(channelResult);
+            return;
+          }
+          default:
+            throw 'Unimplemented';
         }
-        default:
-          throw 'Unimplemented';
-      }
-    };
+      };
 
-    // For the moment, we are only considering directly funded app channels.
-    // Thus, we can directly fetch the channel record, and immediately construct the protocol state from it.
-    // In the future, we can have an App model which collects all the relevant channels for an app channel,
-    // and a Ledger model which stores ledger-specific data (eg. queued requests)
-    const app = await Channel.forId(channels[0], undefined);
+      // For the moment, we are only considering directly funded app channels.
+      // Thus, we can directly fetch the channel record, and immediately construct the protocol state from it.
+      // In the future, we can have an App model which collects all the relevant channels for an app channel,
+      // and a Ledger model which stores ledger-specific data (eg. queued requests)
+      const app = await Store.getChannel(channels[0], tx);
 
-    try {
-      const nextAction = await Application.protocol({app: app.protocolState});
-      // TODO: handleAction might also throw an error.
-      // It would be nice for handleAction to return an Either type, pipe the right values,
-      // and handle the left values with setError
-      await Either.fold(setError, Option.fold(markChannelAsDone, handleAction))(nextAction);
-      await tx.commit();
-    } catch (err) {
-      // FIXME
-      logger.error({err}, 'Error handling action');
-      await setError(err);
-    }
+      if (!app) {
+        setError(new Error('Channel not found'));
+
+        throw 'unreachable';
+      }
+
+      try {
+        const nextAction = await Application.protocol({app});
+        // TODO: handleAction might also throw an error.
+        // It would be nice for handleAction to return an Either type, pipe the right values,
+        // and handle the left values with setError
+        await Either.fold(setError, Option.fold(markChannelAsDone, handleAction))(nextAction);
+      } catch (err) {
+        // FIXME
```

Comment:
Could add a note about what specifically should be fixed

Reply:
https://github.com/statechannels/statechannels/issues/2379